### DISCUSSION TITLE
请问coffee-editor的执行流程是什么？

### DISCUSSION THREAD

**Discussion by 136861535** [eclipse-emfcloud/emfcloud#128] [Coffee Editor]
尊敬的朋友您好，请问coffee-editor的前端界面请求到后端服务的执行流程是什么？请详细解答一下，比如执行到某一个类。

**Comment by planger** [comment]
Hi @136861535,

this is what I got from Google translate:

> Hello dear friends, what is the execution process from the front-end interface of coffee-editor to the back-end service? Please answer in detail, such as executing to a certain class.

Your question is hard to answer in detail as the coffee editor contains many different components, including a form-based editor, a graphical editor, multiple language servers (Java and a custom DSL), code generation and analysis backend.
Architecturally they all work very similarly though:
There is a Theia frontend extension that communicate to a Theia backend, which in turn communicates to an external process (e.g. a language server, graphical language server, code generation executable, etc.), which implements the business logic.

However, pointing to the respective classes can only be done for each of those single components.

In general though, the code of the Theia frontend and backend extensions is located here (the ones ending with -extension):
https://github.com/eclipse-emfcloud/coffee-editor/tree/master/web

The code of the backend logic is located here:
https://github.com/eclipse-emfcloud/coffee-editor/tree/master/backend/plugins

To get an overview about the architecture, I'd also recommend to watch this talk: https://www.youtube.com/watch?v=mEr-v8t2ZQ8

我希望这有帮助

Best wishes,

Philip

**Reply by 136861535** [reply]
你好，我看英文也是用的翻译软件。
我们现在知道前端请求首先请求到下面这个方法中：
![图片](https://user-images.githubusercontent.com/101307516/157791927-3ecf34dd-b0e9-4f7a-822a-2013483cf0b6.png)
在这里我们没法打断点，不知道系统调用后，执行到coffee-editor第一个方法在哪里？请告诉我们，我们好打一个断点进行调试。
非常感谢！！！

**Reply by 136861535** [reply]
上面中的内容是coffee-editor中引用emfcloud-modelserver的包，是class文件，我不知道为什么没有办法打断点调试，所以告诉我们请求进入这里后，往下执行，跳到coffee-editor的第一个方法是什么，在那里我们可以打断点，这样就清楚流程了。谢谢

**Reply by planger** [reply]
If I understand correctly, the question is why the breakpoint doesn't hit.
Afais it should hit as soon as you make a change in the model. But only, if you started the model server from the Eclipse launch configuration in debug mode and the backend connects to that started model server process.

If the model server process isn't started from the launch config in Eclipse, you'd have to attach Eclipse debugging to that process first in order to inspect it, add breakpoints, etc..

**Reply by 136861535** [reply]
断点这个可以不管，我现在主要是想知道，前端的一个请求过来后，coffee-editor中执行的代码在哪里呢？我们需要找到这段代码，以便后续的操作

**Reply by 136861535** [reply]
就是前端发出保存的请求后，在coffee-editor中肯定会执行一段逻辑，那么这段逻辑的代码在哪里呢

**Reply by planger** [reply]
> Breakpoints can be ignored. Now I mainly want to know, where is the code executed in coffee-editor after a request from the front end comes? We need to find this code for subsequent operations
> That is, after the front-end sends a saved request, a piece of logic will definitely be executed in coffee-editor, so where is the code of this logic?

I think you are looking for this method, which will be called on save of a resource:
https://github.com/eclipse-emfcloud/coffee-editor/blob/8456e17b6cc4e75a8a001038622cdedc368bbaf5/backend/plugins/org.eclipse.emfcloud.coffee.modelserver/src/org/eclipse/emfcloud/coffee/modelserver/CoffeeModelResourceManager.java#L99

**Reply by 136861535** [reply]
请问在
https://github.com/eclipse-emfcloud/coffee-editor/blob/8456e17b6cc4e75a8a001038622cdedc368bbaf5/backend/plugins/org.eclipse.emfcloud.coffee.modelserver/src/org/eclipse/emfcloud/coffee/modelserver/CoffeeModelResourceManager.java#L99
文件中的save方法是哪里调用的呢

**Reply by planger** [reply]
In a super class:
https://github.com/eclipse-emfcloud/emfcloud-modelserver/blob/a7e6ea47098ef082545f184755d3cdb779fd9910/bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultModelResourceManager.java#L537

**Reply by 136861535** [reply]
请帮我简单说一下coffee-editor中，每个项目的作用呢(比如："org.eclipse.emfcloud.coffee.modelserver"是服务层，用于执行保存等操作)，我对它一无所知，现在，如图所示，是coffee-editor项目：
![图片](https://user-images.githubusercontent.com/101307516/157834016-c84990a7-5f37-4058-b24a-9f2f8fb1c1c3.png)

**Reply by 136861535** [reply]
我们应该重点关注哪几个项目等等

**Reply by 136861535** [reply]
![图片](https://user-images.githubusercontent.com/101307516/158290266-6d743524-ab7b-4cd0-8d5a-a07ead48ff31.png)
在这段代码中的save方法，我在前端界面修改supperbrewer3000.coffee这个文件时，后端也调用了save方法了，但执行save并没有保存起这个文件，所做的修改没有改变，我看这个方法中，只有一个路径参数，并没有其他所做的参数，这是为什么呢？这是不是只是一个示例，具体的代码 需要我们自己实现呢
